### PR TITLE
[bugfix](vectorization) fix data racing of VDataStreamRecvr

### PR DESCRIPTION
**Pull request by zclllyybb**
# Proposed changes

```
F0505 22:46:41.875746 497986 vdata_stream_recvr.cpp:87] Check failed: _num_remaining_senders == 0 (26 vs. 0) 
*** Check failure stack trace: ***
    @     0x55c8198dc95d  google::LogMessage::Fail()
    @     0x55c8198dee99  google::LogMessage::SendToLog()
    @     0x55c8198dc4c6  google::LogMessage::Flush()
    @     0x55c8198df509  google::LogMessageFatal::~LogMessageFatal()
    @     0x55c81814cccb  doris::vectorized::VDataStreamRecvr::SenderQueue::_inner_get_batch()
    @     0x55c81815fa74  doris::vectorized::VDataStreamRecvr::PipSenderQueue::get_batch()
    @     0x55c818155072  doris::vectorized::VDataStreamRecvr::get_next()
    @     0x55c808360620  doris::vectorized::VExchangeNode::get_next()
    @     0x55c80071a114  doris::ExecNode::pull()
    @     0x55c818e68e8d  std::__invoke_impl<>()
    @     0x55c818e68c30  std::__invoke<>()
    @     0x55c818e68b67  _ZNSt5_BindIFMN5doris8ExecNodeEFNS0_6StatusEPNS0_12RuntimeStateEPNS0_10vectorized5BlockEPbEPNS5_13VExchangeNodeESt12_PlaceholderILi1EESD_ILi2EESD_ILi3EEEE6__callIS2_JOS4_OS7_OS8_EJLm0ELm1ELm2ELm3EEEET_OSt5tupleIJDpT0_EESt12_Index_tupleIJXspT1_EEE
    @     0x55c818e688f8  std::_Bind<>::operator()<>()
    @     0x55c818e687a8  std::__invoke_impl<>()
    @     0x55c818e68718  _ZSt10__invoke_rIN5doris6StatusERSt5_BindIFMNS0_8ExecNodeEFS1_PNS0_12RuntimeStateEPNS0_10vectorized5BlockEPbEPNS6_13VExchangeNodeESt12_PlaceholderILi1EESE_ILi2EESE_ILi3EEEEJS5_S8_S9_EENSt9enable_ifIX16is_invocable_r_vIT_T0_DpT1_EESM_E4typeEOSN_DpOSO_
    @     0x55c818e68368  std::_Function_handler<>::_M_invoke()
    @     0x55c800719fff  std::function<>::operator()()
    @     0x55c80070e18a  doris::ExecNode::get_next_after_projects()
    @     0x55c818e655f3  doris::pipeline::SourceOperator<>::get_block()
    @     0x55c818e05c3f  doris::pipeline::PipelineTask::execute()
    @     0x55c818e2429a  doris::pipeline::TaskScheduler::_do_work()
```

## Problem summary

the check of emptiness should be locked in the same lock of real use to ensure correctness. usage before is defective. now fixed it.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [x] Is this PR support rollback (If NO, please explain WHY)

## Further comments

the lock may could be split for the usage of the specific object.

### CODE REVIEW

**Comment by zclllyybb** [comment]
run buildall

File: be/src/vec/runtime/vdata_stream_recvr.cpp [hunk 1]
```diff
@@ -161,9 +162,11 @@ void VDataStreamRecvr::SenderQueue::add_block(const PBlock& pblock, int be_numbe
     COUNTER_UPDATE(_recvr->_deserialize_row_batch_timer, deserialize_time);
     COUNTER_UPDATE(_recvr->_decompress_timer, block->get_decompress_time());
     COUNTER_UPDATE(_recvr->_decompress_bytes, block->get_decompressed_bytes());
-
-    _block_queue.emplace_back(std::move(block), block_byte_size);
-    _update_block_queue_empty();
+    {
+        std::lock_guard<std::mutex> l(_lock);
+        _block_queue.emplace_back(std::move(block), block_byte_size);
+        _update_block_queue_empty_in_lock();
```
Comment (BiteTheDDDDt) [inline be/src/vec/runtime/vdata_stream_recvr.cpp:168]: Maybe we can wrap this piece of code into a function
```
lock
push block to queue
update emtpy flag
```

Comment (zclllyybb) [reply be/src/vec/runtime/vdata_stream_recvr.cpp:168]: that‘s right. will do.

File: be/src/vec/runtime/vdata_stream_recvr.cpp [hunk 2]
```diff
@@ -83,19 +82,20 @@ Status VDataStreamRecvr::SenderQueue::_inner_get_batch(Block* block, bool* eos)
         return Status::Cancelled("Cancelled");
     }
 
-    if (_block_queue.empty()) {
+    std::unique_lock<std::mutex> l(_lock);
```
Comment (HappenLee) [inline be/src/vec/runtime/vdata_stream_recvr.cpp:85]: should not only add a mutex here. please review the whloe logic to find problem

**Comment by zclllyybb** [comment]
run buildall

**Pull request closed without merging**